File: yurt-jdbc/src/main/java/com/github/liulus/yurt/jdbc/SQL.java
package com.github.liulus.yurt.jdbc;

import java.util.Arrays;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/9
 */
public class SQL {
    private final SQLStatement sql = new SQLStatement();


    public SQL UPDATE(String table) {
        sql.statementType = SQLStatement.StatementType.UPDATE;
        sql.tables.add(table);
        return this;
    }

    public SQL SET(String sets) {
        sql.sets.add(sets);
        return this;
    }

    public SQL SET(String... sets) {
        sql.sets.addAll(Arrays.asList(sets));
        return this;
    }

    public SQL INSERT_INTO(String tableName) {
        sql.statementType = SQLStatement.StatementType.INSERT;
        sql.tables.add(tableName);
        return this;
    }

    public SQL VALUES(String columns, String values) {
        sql.columns.add(columns);
        sql.values.add(values);
        return this;
    }

    public SQL INTO_COLUMNS(String... columns) {
        sql.columns.addAll(Arrays.asList(columns));
        return this;
    }

    public SQL INTO_VALUES(String... values) {
        sql.values.addAll(Arrays.asList(values));
        return this;
    }

    public SQL SELECT(String columns) {
        sql.statementType = SQLStatement.StatementType.SELECT;
        sql.select.add(columns);
        return this;
    }

    public SQL SELECT(String... columns) {
        sql.statementType = SQLStatement.StatementType.SELECT;
        sql.select.addAll(Arrays.asList(columns));
        return this;
    }

    public SQL SELECT_DISTINCT(String columns) {
        sql.distinct = true;
        SELECT(columns);
        return this;
    }

    public SQL SELECT_DISTINCT(String... columns) {
        sql.distinct = true;
        SELECT(columns);
        return this;
    }

    public SQL DELETE_FROM(String table) {
        sql.statementType = SQLStatement.StatementType.DELETE;
        sql.tables.add(table);
        return this;
    }

    public SQL FROM(String table) {
        sql.tables.add(table);
        return this;
    }

    public SQL FROM(String... tables) {
        sql.tables.addAll(Arrays.asList(tables));
        return this;
    }

    public SQL JOIN(String join) {
        sql.join.add(join);
        return this;
    }

    public SQL JOIN(String... joins) {
        sql.join.addAll(Arrays.asList(joins));
        return this;
    }

    public SQL INNER_JOIN(String join) {
        sql.innerJoin.add(join);
        return this;
    }

    public SQL INNER_JOIN(String... joins) {
        sql.innerJoin.addAll(Arrays.asList(joins));
        return this;
    }

    public SQL LEFT_OUTER_JOIN(String join) {
        sql.leftOuterJoin.add(join);
        return this;
    }

    public SQL LEFT_OUTER_JOIN(String... joins) {
        sql.leftOuterJoin.addAll(Arrays.asList(joins));
        return this;
    }

    public SQL RIGHT_OUTER_JOIN(String join) {
        sql.rightOuterJoin.add(join);
        return this;
    }

    public SQL RIGHT_OUTER_JOIN(String... joins) {
        sql.rightOuterJoin.addAll(Arrays.asList(joins));
        return this;
    }

    public SQL OUTER_JOIN(String join) {
        sql.outerJoin.add(join);
        return this;
    }

    public SQL OUTER_JOIN(String... joins) {
        sql.outerJoin.addAll(Arrays.asList(joins));
        return this;
    }

    public SQL WHERE(String conditions) {
        sql.where.add(conditions);
        sql.lastList = sql.where;
        return this;
    }

    public SQL WHERE(String... conditions) {
        sql.where.addAll(Arrays.asList(conditions));
        sql.lastList = sql.where;
        return this;
    }

    public SQL OR() {
        sql.lastList.add(SQLStatement.OR);
        return this;
    }

    public SQL AND() {
        sql.lastList.add(SQLStatement.AND);
        return this;
    }

    public SQL GROUP_BY(String columns) {
        sql.groupBy.add(columns);
        return this;
    }

    public SQL GROUP_BY(String... columns) {
        sql.groupBy.addAll(Arrays.asList(columns));
        return this;
    }

    public SQL HAVING(String conditions) {
        sql.having.add(conditions);
        sql.lastList = sql.having;
        return this;
    }

    public SQL HAVING(String... conditions) {
        sql.having.addAll(Arrays.asList(conditions));
        sql.lastList = sql.having;
        return this;
    }

    public SQL ORDER_BY(String columns) {
        sql.orderBy.add(columns);
        return this;
    }

    public SQL ORDER_BY(String... columns) {
        sql.orderBy.addAll(Arrays.asList(columns));
        return this;
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        return sql.sql(sb);
    }

    public String toCountSql() {
        sql.statementType = SQLStatement.StatementType.COUNT;
        String result = toString();
        sql.statementType = SQLStatement.StatementType.SELECT;
        return result;
    }
}

File: yurt-data/yurt-data-jdbc/src/main/kotlin/com/github/liulus/yurt/jdbc/SQL.kt
package com.github.liulus.yurt.jdbc

/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/11/2
 */
class SQL {

    private val sql = SQLStatement()

    fun UPDATE(table: String): SQL {
        sql.statementType = SQLStatement.StatementType.UPDATE
        sql.tables.add(table)
        return this
    }

    fun SET(sets: String): SQL {
        sql.sets.add(sets)
        return this
    }

    fun SET(vararg sets: String): SQL {
        sql.sets.addAll(listOf(*sets))
        return this
    }

    fun INSERT_INTO(tableName: String): SQL {
        sql.statementType = SQLStatement.StatementType.INSERT
        sql.tables.add(tableName)
        return this
    }

    fun VALUES(columns: String, values: String): SQL {
        sql.columns.add(columns)
        sql.values.add(values)
        return this
    }

    fun INTO_COLUMNS(vararg columns: String): SQL {
        sql.columns.addAll(listOf(*columns))
        return this
    }

    fun INTO_VALUES(vararg values: String): SQL {
        sql.values.addAll(listOf(*values))
        return this
    }

    fun SELECT(columns: String): SQL {
        sql.statementType = SQLStatement.StatementType.SELECT
        sql.select.add(columns)
        return this
    }

    fun SELECT(vararg columns: String): SQL {
        sql.statementType = SQLStatement.StatementType.SELECT
        sql.select.addAll(listOf(*columns))
        return this
    }

    fun SELECT_DISTINCT(columns: String): SQL {
        sql.distinct = true
        SELECT(columns)
        return this
    }

    fun SELECT_DISTINCT(vararg columns: String): SQL {
        sql.distinct = true
        SELECT(*columns)
        return this
    }

    fun DELETE_FROM(table: String): SQL {
        sql.statementType = SQLStatement.StatementType.DELETE
        sql.tables.add(table)
        return this
    }

    fun FROM(table: String): SQL {
        sql.tables.add(table)
        return this
    }


    fun FROM(vararg tables: String): SQL {
        sql.tables.addAll(listOf(*tables))
        return this
    }

    fun JOIN(join: String): SQL {
        sql.join.add(join)
        return this
    }

    fun JOIN(vararg joins: String): SQL {
        sql.join.addAll(listOf(*joins))
        return this
    }

    fun INNER_JOIN(join: String): SQL {
        sql.innerJoin.add(join)
        return this
    }

    fun INNER_JOIN(vararg joins: String): SQL {
        sql.innerJoin.addAll(listOf(*joins))
        return this
    }

    fun LEFT_OUTER_JOIN(join: String): SQL {
        sql.leftOuterJoin.add(join)
        return this
    }

    fun LEFT_OUTER_JOIN(vararg joins: String): SQL {
        sql.leftOuterJoin.addAll(listOf(*joins))
        return this
    }

    fun RIGHT_OUTER_JOIN(join: String): SQL {
        sql.rightOuterJoin.add(join)
        return this
    }

    fun RIGHT_OUTER_JOIN(vararg joins: String): SQL {
        sql.rightOuterJoin.addAll(listOf(*joins))
        return this
    }

    fun OUTER_JOIN(join: String): SQL {
        sql.outerJoin.add(join)
        return this
    }

    fun OUTER_JOIN(vararg joins: String): SQL {
        sql.outerJoin.addAll(listOf(*joins))
        return this
    }

    fun WHERE(conditions: String): SQL {
        sql.where.add(conditions)
        sql.lastList = sql.where
        return this
    }

    fun WHERE(vararg conditions: String): SQL {
        sql.where.addAll(listOf(*conditions))
        sql.lastList = sql.where
        return this
    }

    fun OR(): SQL {
        sql.lastList.add(SQLStatement.OR)
        return this
    }

    fun AND(): SQL {
        sql.lastList.add(SQLStatement.AND)
        return this
    }

    fun GROUP_BY(columns: String): SQL {
        sql.groupBy.add(columns)
        return this
    }

    fun GROUP_BY(vararg columns: String): SQL {
        sql.groupBy.addAll(listOf(*columns))
        return this
    }

    fun HAVING(conditions: String): SQL {
        sql.having.add(conditions)
        sql.lastList = sql.having
        return this
    }

    fun HAVING(vararg conditions: String): SQL {
        sql.having.addAll(listOf(*conditions))
        sql.lastList = sql.having
        return this
    }

    fun ORDER_BY(columns: String): SQL {
        sql.orderBy.add(columns)
        return this
    }

    fun ORDER_BY(vararg columns: String): SQL {
        sql.orderBy.addAll(listOf(*columns))
        return this
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sql.sql(sb)
        return sb.toString()
    }

    fun toCountSql(): String {
        sql.statementType = SQLStatement.StatementType.COUNT
        val result = toString()
        sql.statementType = SQLStatement.StatementType.SELECT
        return result
    }

}


